src/coreclr/jit/optimizer.cpp

```diff
@@ -9335,6 +9337,262 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
     return true;
 }
 
+//-----------------------------------------------------------------------------
+//  FindCompareChain:  Check if the given condition is a compare chain.
+//
+// Arguments:
+//      condition:        Condition to check.
+//      isTestCondition:  Returns true if condition is a EQ/NE(AND(...),0) but is not a compare chain.
+//
+//  Returns:
+//      true if chain optimization is a compare chain.
+//
+//  Assumptions:
+//      m_b1 and m_b2 are set on entry.
+//
+
+inline bool OptBoolsDsc::FindCompareChain(GenTree* condition, bool* isTestCondition)
+{
+    GenTree* condOp1 = condition->gtGetOp1();
+    GenTree* condOp2 = condition->gtGetOp2();
+
+    *isTestCondition = false;
+
+    if (condition->OperIs(GT_EQ, GT_NE) && condOp2->IsIntegralConst() && condOp2->AsIntCon()->IconValue() == 0)
+    {
+        // Found a test condition. Does it contain a compare chain?
+
+        if (condOp1->OperIs(GT_AND, GT_OR) && varTypeIsIntegralOrI(condOp1->gtGetOp1()) &&
+            varTypeIsIntegralOrI(condOp1->gtGetOp2()))
```

Comment:
```suggestion
        if (condOp1->OperIs(GT_AND, GT_OR))
```

These should be always true -- if the intention is to check for candidates that lowering can convert to ccmp then I think you need those checks on one of the operands of the relops below (e.g. `varTypeIsIntegralOrI(condOp1->gtGetOp1()->gtGetOp1())`).